First Implementation of Spell feature

davidgoering opened the pull request:
Fairly straight forward.
Things to discuss:
- Do we want to be dependant on an external interface or do we want to keep a local json file as the underlying data?
- I wasn't quite sure how to manage a fuzzy search on the dictionary key. At the moment it will return the first spell it finds that matches (problem case ie "fireball" )

davidgoering committed:
first prototype for spell feature

davidgoering committed:
Added missing files.

davidgoering committed:
Removed unnecessary variable.

cameronpenner commented on DiscordBot/SpellSearch.cs:
```diff
@@ -0,0 +1,43 @@
+        {
+            JsonSerializer serializer = new JsonSerializer();
+            using(var client = new WebClient()){
+                var json = client.DownloadString("https://raw.githubusercontent.com/tadzik/5e-spells/master/spells.json");
```
if we store the dictionary at class level, then we won't have to download it and convert it every time we want to check a spell. we could make it static :( and only build the dictionary if it hasn't been already. Or we could make SpellSearch a singleton :( or plug in a simple IoC container.

davidgoering replied:
Good point, I will definitely move it out so that it is only done once.

cameronpenner commented on DiscordBot/SpellSearch.cs:
```diff
@@ -0,0 +1,43 @@
+            MyStringBuilder.AppendLine("> **School:** " + result.Value.school);
+            return MyStringBuilder.ToString();
+            
+
```
I'm pretty picky with my whitespace, we probably could get rid of line 16, 29, 30, and 40 in this file. Check out CodeMaid for visual studio, it's great for cleaning up simple formatting like this.

davidgoering replied:
Yeah like I would know the shortcut for auto formatting in IntelliJ in a heart beat... but thanks for the tip on the extension will install it right now ;)

cameronpenner commented on DiscordBot/SpellSearch.cs:
```diff
@@ -0,0 +1,43 @@
+            Dictionary <string, Spell> spells = map();
+            KeyValuePair<string, Spell> result = spells.FirstOrDefault(kvp => kvp.Key.ToLower().Contains(query.ToLower()));
+            StringBuilder MyStringBuilder = new StringBuilder();
+            MyStringBuilder.AppendLine("> **Spell:** *" + result.Key + "*");
```
What does the "> " do here? will that be visible or is it a formatting thing?

davidgoering replied:
Well it was supposed to based on this: https://sourceforge.net/p/discord/wiki/markdown_syntax/#md_ex_bq

But it didn't work, so it shows up... but I think it looks better with the ">" than without... if we can figure out how to format it in a better way I am all ears.

davidgoering committed:
- Added monster Search.
- Added Singleton implementation for the Search classes (one time parsing of json)
- General cleanup & nicer printing.

cameronpenner commented on DiscordBot/MonsterSearch.cs:
```diff
@@ -0,0 +1,65 @@
+        {
+        }
+
+        public static MonsterSearch Instance
```
What do you think of making a singleton class and inheriting from that? less copy pasta code.
